sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamEnumerableConverter.java

```diff
@@ -125,6 +162,55 @@ private static PipelineResult run(
     return result;
   }
 
+  private static PipelineResult limitRun(
+      PipelineOptions options,
+      BeamRelNode node,
+      DoFn<Row, KV<String, Row>> collectDoFn,
+      DoFn<KV<String, Row>, Void> limitCounterDoFn,
+      LimitStateVar limitStateVar) {
+    ExecutorService pool = Executors.newFixedThreadPool(1);
+
+    options.as(DirectOptions.class).setBlockOnRun(false);
+    Pipeline pipeline = Pipeline.create(options);
+    PCollectionTuple.empty(pipeline)
+        .apply(node.toPTransform())
+        .apply(ParDo.of(collectDoFn))
+        .apply(ParDo.of(limitCounterDoFn));
+
+    PipelineResult result = pipeline.run();
+
+    pool.execute(
+        new Runnable() {
+          @Override
+          public void run() {
+            // wait every one second to check state of PipelineResult. If returned value
+            // is null, or return state is non termination state, it means pipeline is still
+            // running and within loop limit counter state will be checked.
+            State state;
+            while (true) {
+              state = result.waitUntilFinish(Duration.standardSeconds(1));
+              if (state != null && state.isTerminal()) {
+                break;
+              }
+              // If state is null, or state is not null and indicates pipeline is not terminated
+              // yet, check continue checking the limit var.
+              try {
+                if (limitStateVar.isReached()) {
```

Comment:
If you did something like `values.size() >= limitCount` here you could replace `LimitCanceller` and `LimitCollector` with `Collector`.

Reply:
I think we have a very different model as to how this will work in other runners. The runners shouldn't need to be aware of the limit, as that would be applied client side.